system/trace/trace_dump.c

```diff
@@ -0,0 +1,643 @@
+    {
+      return tctx->name;
+    }
+
+  return "<noname>";
+}
+
+/****************************************************************************
+ * Name: trace_dump_header
+ ****************************************************************************/
+
+static void trace_dump_header(FILE *out,
+                              FAR struct note_common_s *note,
+                              FAR struct trace_dump_context_s *ctx)
+{
+  pid_t pid;
+  uint32_t systime = note->nc_systime[0] +
+                     (note->nc_systime[1] << 8) +
+                     (note->nc_systime[2] << 16) +
+                     (note->nc_systime[3] << 24);
+#ifdef CONFIG_SMP
+  int cpu = note->nc_cpu;
+#else
+  int cpu = 0;
+#endif
+
+  pid = ctx->cpu[cpu].current_pid;
+
+  fprintf(out, "%8s-%-3u [%d] %3u.%09u: ",
+          get_task_name(pid, ctx), get_pid(pid), cpu,
+          systime / (1000 * 1000 / CONFIG_USEC_PER_TICK),
+          (systime % (1000 * 1000 / CONFIG_USEC_PER_TICK))
+            * CONFIG_USEC_PER_TICK * 1000);
+}
+
+/****************************************************************************
+ * Name: trace_dump_sched_switch
+ ****************************************************************************/
+
+static void trace_dump_sched_switch(FILE *out,
+                                    FAR struct note_common_s *note,
+                                    FAR struct trace_dump_context_s *ctx)
+{
+  FAR struct trace_dump_cpu_context_s *cctx;
+  pid_t current_pid;
+  pid_t next_pid;
+#ifdef CONFIG_SMP
+  int cpu = note->nc_cpu;
+#else
+  int cpu = 0;
+#endif
+
+  cctx = &ctx->cpu[cpu];
+  current_pid = cctx->current_pid;
+  next_pid = cctx->next_pid;
+
+  fprintf(out, "sched_switch: "
+               "prev_comm=%s prev_pid=%u prev_state=%c ==> "
+               "next_comm=%s next_pid=%u\n",
+          get_task_name(current_pid, ctx), get_pid(current_pid),
+          get_task_state(cctx->current_state),
+          get_task_name(next_pid, ctx), get_pid(next_pid));
+
+  cctx->current_pid = cctx->next_pid;
+  cctx->pendingswitch = false;
+}
+
+/****************************************************************************
+ * Name: trace_dump_one
+ ****************************************************************************/
+
+static int trace_dump_one(FILE *out,
+                          FAR uint8_t *p,
+                          FAR struct trace_dump_context_s *ctx)
+{
+  FAR struct note_common_s *note = (FAR struct note_common_s *)p;
+  FAR struct trace_dump_cpu_context_s *cctx;
+  pid_t pid;
+#ifdef CONFIG_SMP
+  int cpu = note->nc_cpu;
+#else
+  int cpu = 0;
+#endif
+
+  cctx = &ctx->cpu[cpu];
+  pid = note->nc_pid[0] + (note->nc_pid[1] << 8);
+
+  if (note->nc_type != NOTE_START &&
+      note->nc_type != NOTE_STOP &&
+      note->nc_type != NOTE_RESUME
+#ifdef CONFIG_SMP
+      && !(note->nc_type >= NOTE_CPU_START &&
+           note->nc_type <= NOTE_CPU_RESUMED)
+#endif
+     )
+    {
+      cctx->current_pid = pid;
```

Comment:
It is needed because cctx->current_pid isn't updated until the first context switch in the trace.
The trace data begins in incomplete state and the beginning of the trace dump also shows incomplete result. This line fixes such situation though it is not perfect.